src/core/qtlibtorrent/qbtsession.cpp

```diff
@@ -1276,72 +1276,78 @@ void QBtSession::loadTorrentTempData(QTorrentHandle &h, QString savePath, bool m
-    QList<QUrl> new_trackers = misc::magnetUriToTrackers(magnet_uri);
-    bool trackers_added = false;
-    foreach (const QUrl& new_tracker, new_trackers) {
+    QStringList trackers = misc::magnetUriToTrackers(magnet_uri);
+    QStringList urlSeeds = misc::magnetUriToUrlSeeds(magnet_uri);
+    mergeTorrents_impl(h, trackers, urlSeeds);
+}
+
+void QBtSession::mergeTorrents(const QTorrentHandle &h, const boost::intrusive_ptr<torrent_info> t) {
+    QStringList trackers;
+    QStringList urlSeeds;
+
+    foreach (const announce_entry& newTracker, t->trackers())
+        trackers.append(misc::toQStringU(newTracker.url));
+
+    foreach (const web_seed_entry& newUrlSeed, t->web_seeds())
+        urlSeeds.append(misc::toQStringU(newUrlSeed.url));
+
+    mergeTorrents_impl(h, trackers, urlSeeds);
+}
+
+void QBtSession::mergeTorrents_impl(const QTorrentHandle &h, const QStringList &trackers, const QStringList &urlSeeds)
+{
+    if (!h.is_valid())
+        return;
+
+    QString hash = h.hash();
+    QString name = h.name();
+    const std::vector<announce_entry> existingTrackers = h.trackers();
+    const QStringList existingUrlSeeds = h.url_seeds();
+
+    foreach (const QString &tracker, trackers) {
+        QUrl trackerUrl(tracker);
         bool found = false;
-        std::vector<announce_entry> existing_trackers = h_ex.trackers();
-        foreach (const announce_entry& existing_tracker, existing_trackers) {
-            if (new_tracker == QUrl(existing_tracker.url.c_str())) {
+
+        foreach (const announce_entry &existingTracker, existingTrackers) {
+            QUrl existingTrackerUrl(misc::toQStringU(existingTracker.url));
+            if (trackerUrl == existingTrackerUrl) {
                 found = true;
                 break;
             }
         }
 
         if (!found) {
-            h_ex.add_tracker(announce_entry(new_tracker.toString().toStdString()));
-            trackers_added = true;
+            h.add_tracker(announce_entry(tracker.toUtf8().constData()));
+            emit trackerAdded(tracker, hash);
+            Logger::instance()->addMessage(tr("Tracker '%1' was added to torrent '%2'").arg(tracker).arg(name));
         }
     }
-    if (trackers_added)
-        Logger::instance()->addMessage(tr("Note: new trackers were added to the existing torrent."));
-}
 
-void QBtSession::mergeTorrents(QTorrentHandle &h_ex, boost::intrusive_ptr<torrent_info> t) {
-    // Check if the torrent contains trackers or url seeds we don't know about
-    // and add them
-    if (!h_ex.is_valid()) return;
-    std::vector<announce_entry> existing_trackers = h_ex.trackers();
-    std::vector<announce_entry> new_trackers = t->trackers();
-    bool trackers_added = false;
-    foreach (const announce_entry& new_tracker, new_trackers) {
-        std::string new_tracker_url = new_tracker.url;
-        // Check if existing torrent has this tracker
+    if (existingTrackers.empty() && !h.trackers().empty())
+        emit trackerRemoved("", hash);
```

Comment:
I haven't got time now to fully understand your code. But I can indicate another design mistake that has an effect on it.
Your widget should remain abstract from certain types of data (for example, from TorrentHandle, like you want). But it directly interacts with the session and also controls the filter. This is wrong. The filter needs to interact with the session and the widget is just to reflect its state and to provide the user with a mechanism of interaction with it.
Then you can pass TorrentHandle (to filter) in the signal, and it (filter) will fulfill all of transformative operations and send their signals to the widget.
P.S. I use here the word 'filter', because something like you have. Don't pay any attention to it. The General point is this: there is some component that contains the necessary logic and you have some visual component (widget), which provides a GUI for the first component. The first should not depend on how it is used.